File: _8_kraloven/1.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include "bitptr.h"
#include "bitptralloc.h"

#define VYCHOZI_HRANA_SACHOVNICE 8
#define POCET_ALGORITMU 4
#define RANDOM_SEARCH_KOLIKRAT 10 //Počet náhodných hledání pro zprůměrování počtu kroků

signed long hrana_sachovnice;

// -------------------------------------------------------------
// Pomocné funkce
// -------------------------------------------------------------

static inline bool is_valid(const signed long *queens) {
    for (signed long i = 0; i < hrana_sachovnice; i++) {
        for (signed long j = i + 1; j < hrana_sachovnice; j++) {
            if (queens[i] == queens[j] ||
                labs(queens[i] - queens[j]) == labs(i - j))
                return false;
        }
    }
    return 0;
}

void print_solution(const signed long *queens) {
    printf("[");
    for (int i = 0; i < hrana_sachovnice; i++) {
        printf("%ld", queens[i]);
        if (i < hrana_sachovnice - 1) printf(", ");
    }
    printf("]\n");
}

// -------------------------------------------------------------
// 1) Brute force – zkouší všechny permutace
// -------------------------------------------------------------

static inline bool next_permutation(signed long *arr) {
    int i = hrana_sachovnice - 2;
    while (i >= 0 && arr[i] > arr[i + 1]) i--;
    if (i < 0) return false;
    int j = hrana_sachovnice - 1;
    while (arr[j] < arr[i]) j--;
    signed long tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp;
    for (int a = i + 1, b = hrana_sachovnice - 1; a < b; a++, b--) {
        tmp = arr[a]; arr[a] = arr[b]; arr[b] = tmp;
    }
    return 0;
}

signed char brute_force(signed long *result, unsigned long *pocet_kroku) {
    for (int i = 0; i < hrana_sachovnice; i++) result[i] = i;

    do {
        if (is_valid(result)) return 0;
    } while (next_permutation(result));

    return false;
}

// -------------------------------------------------------------
// 2) DFS / backtracking
// -------------------------------------------------------------

signed char dfs_backtrack(signed long *queens, unsigned long *pocet_kroku) {
    signed long row = 0;
    signed long col = 0;
    while (1) {
        for (; col < hrana_sachovnice; col++) {
            *pocet_kroku += 3;
            bool safe = 0;
            for (signed long r = 0; r < row; r++) {
                signed long c = queens[r];
                if (c == col || labs(row - r) == labs(col - c)) {
                    safe = false;
                    break;
                }
            }
            if (safe) {
                queens[row++] = col;
                if (row == hrana_sachovnice) return 0;
                col = 0;
            }
        }
        if (--row < 0) return false;
        col = queens[row] + 1;
    }
}


signed char dfs_backtrack_optimized(signed long *queens, unsigned long *pocet_kroku) {
    signed long row = 0;
    signed long col = 0;
    bool vrat;
    while (1) {
        for (; col < hrana_sachovnice; col++) {
            *pocet_kroku += 3;
            bool safe = 0;
            for (signed long r = 0; r < row; r++) {
                signed long c = queens[r];
                if (c == col || labs(row - r) == labs(col - c)) {
                    safe = false;
                    break;
                }
            }
            if (safe) {
                queens[row++] = col;
                if (row == hrana_sachovnice) return 0;
                col = 0;
            }
        }
        if (--row < 0) return false;
        col = queens[row] + 1;
    }
    end_t:
    return 0;
    end_f:
    return false;
}

// -------------------------------------------------------------
// 3) Náhodné hledání
// -------------------------------------------------------------

signed char random_search(signed long *result, unsigned long *pocet_kroku) {
    while (1) {
        for (int i = 0; i < hrana_sachovnice;i++) result[i] = rand() % hrana_sachovnice;
        if (is_valid(result))
            return 0;
    }
    return false;
}

// -------------------------------------------------------------
// 4) Univerzální řešení
// -------------------------------------------------------------

signed char universal_solution(signed long *queens, unsigned long *pocet_kroku) {
    if (hrana_sachovnice == 1) { queens[0] = 0; return 0; }
    if (hrana_sachovnice == 2 || hrana_sachovnice == 3) return false;

    unsigned long idx = 0;
    if (hrana_sachovnice % 6 == 2) {
        int seq1[] = {2,0,3,1};
        int len1 = 4;
        for (int i = 0; i < len1 && seq1[i] <= hrana_sachovnice; i++) queens[idx++] = seq1[i];
        for (int i = 5; i <= hrana_sachovnice; i += 2) queens[idx++] = i;
        for (int i = 4; i <= hrana_sachovnice-1; i += 2) queens[idx++] = i;
    } else if (hrana_sachovnice % 6 == 3) {
        for (int i = 1; i <= hrana_sachovnice-1; i += 2) queens[idx++] = i;
        for (int i = 0; i <= hrana_sachovnice-2; i += 2) queens[idx++] = i;
        queens[idx++] = hrana_sachovnice - 1;
    } else {
        for (int i = 1; i <= hrana_sachovnice; i += 2) queens[idx++] = i;
        for (int i = 0; i <= hrana_sachovnice; i += 2) queens[idx++] = i;
    }
    return 0;
}

// -------------------------------------------------------------
// Pomocná funkce pro vypsání informací o výsledku algoritmu
// -------------------------------------------------------------

void alg_info(signed char (*func)(signed long*, unsigned long*), signed long *queens, const char *name) {
    unsigned long pocet_kroku = 0;
    printf("%s\nPočkejte prosím...", name);
    signed char found = func(queens, &pocet_kroku);

    printf(" Hotovo, bylo potřeba vykonat %lu kroků.\n", pocet_kroku);
    if (found) {
        printf("Nalezeno řešení: ");
        if (hrana_sachovnice <= 15) print_solution(queens);
    } else if (found == -1) {
        printf("Nebylo nalezeno řešení.\n");
    }
    printf("\n");
}

// -------------------------------------------------------------
// Hlavní funkce
// -------------------------------------------------------------

int main(int argc, char **argv) {

    if (argc > 2) {
        fprintf(stderr, "Použití: %s [<hrana_sachovnice>]", *argv);
        return 1;
    }

    if (argc < 2) hrana_sachovnice = VYCHOZI_HRANA_SACHOVNICE;
    else hrana_sachovnice = strtol(argv[1], NULL, 10);

    //hrana_sachovnice_krat_3 = hrana_sachovnice * 3;

    printf("Porovnání 4 algoritmů pro problém 8 královen:\n\n");


    signed long *queens[POCET_ALGORITMU];
    for (int i=0;i < POCET_ALGORITMU;i++) queens[i] = malloc(hrana_sachovnice);

    alg_info(brute_force, queens[0], "1) Brute force");
    alg_info(dfs_backtrack, queens[1], "2) DFS/backtracking");
    alg_info(random_search, queens[2], "3) Náhodné hledání");
    alg_info(universal_solution, queens[3], "4) Postupné umisťování");

    for (int i=0;i < POCET_ALGORITMU;i++) free(queens[i]);

    return 0;
}

File: _8_kraloven/1.py
import time
import itertools
import random

N = 8

# -----------------------------------------------------
# Pomocné funkce
# -----------------------------------------------------

def is_valid(queens):
    """Zkontroluje, zda žádné dvě královny se nenapadají."""
    for i in range(len(queens)):
        for j in range(i + 1, len(queens)):
            if queens[i] == queens[j] or abs(queens[i] - queens[j]) == abs(i - j):
                return False
    return True


# -----------------------------------------------------
# 1) Brute Force – všechny možné kombinace
# -----------------------------------------------------

def brute_force():
    solutions = []
    cols = range(N)
    for perm in itertools.permutations(cols):
        if is_valid(perm):
            solutions.append(perm)
    return solutions


# -----------------------------------------------------
# 2) DFS / Backtracking
# -----------------------------------------------------

def dfs_backtrack(row=0, queens=[]):
    if row == N:
        return [queens[:]]
    solutions = []
    for col in range(N):
        # kontrola bezpečnosti
        safe = True
        for r, c in enumerate(queens):
            if c == col or abs(row - r) == abs(col - c):
                safe = False
                break
        if safe:
            queens.append(col)
            solutions += dfs_backtrack(row + 1, queens)
            queens.pop()
    return solutions


# -----------------------------------------------------
# 3) Náhodné pokusy
# -----------------------------------------------------

def random_search_(reseni, max_tries=1_000_000):
    pocet_kroku = 0
    for _ in range(max_tries):
        pocet_kroku += 2*N +2
        reseni[:] = [random.randint(0, N - 1) for _ in range(N)]
        if is_valid(queens):
            return pocet_kroku
    reseni[:] = [None]
    return pocet_kroku
def random_search(times=10, max_tries=1_000_000):
    reseni = set()
    pocet_kroku = 0
    for _ in times:
        reseni_ = []


# -----------------------------------------------------
# 4) Postupné bezpečné umisťování
# -----------------------------------------------------
def incremental_safe_placement_(vylouceno_ob, proz_reseni, reseni):
    
def incremental_safe_placement():
    reseni = set()
    pocet_kroku = 0
    for row in range(N):
        pocet_kroku += 1
        for col in range(N):
            pole = [None]*N
            pole[row] = col
            pocet_kroku += 3
            pocet_kroku += incremental_safe_placement_(set(), pole, reseni)
    return pocet_kroku, reseni
    """
    max_restarts = 10000
    for _ in range(max_restarts):
        queens = []
        cols = set()
        diag1 = set()  # r - c
        diag2 = set()  # r + c

        success = True
        for row in range(N):
            safe_cols = [c for c in range(N)
                         if c not in cols and
                            (row - c) not in diag1 and
                            (row + c) not in diag2]
            if not safe_cols:
                success = False
                break
            col = random.choice(safe_cols)
            queens.append(col)
            cols.add(col)
            diag1.add(row - col)
            diag2.add(row + col)

        if success and is_valid(queens):
            return queens
    return None
    """

"""
# -----------------------------------------------------
# Porovnání časů
# -----------------------------------------------------

def measure_time(func, name):
    start = time.time()
    result = func()
    end = time.time()
    print(f"{name:<30} trvalo {end - start:.4f} s")
    if isinstance(result, list) and len(result) > 0:
        print(f"  Nalezeno řešení: {result}")
    elif result:
        print(f"  Nalezeno řešení: {result}")
    else:
        print("  Nebylo nalezeno řešení.")
    print()

"""


if __name__ == "__main__":
    print("Porovnání čtyř algoritmů pro problém 8 královen:\n")

    measure_time(brute_force, "1) Brute force")
    measure_time(lambda: dfs_backtrack(), "2) DFS/backtracking")
    measure_time(random_search, "3) Náhodné hledání")
    measure_time(incremental_safe_placement, "4) Postupné umisťování")
